### ISSUE TITLE
[data grid] Potential Memory Leak in the DataGrid component

### ISSUE THREAD

**Issue by marcusguay** [mui/mui-x#15459]
### Steps to reproduce

Steps:
1. Open this link to live example: https://codesandbox.io/p/devbox/vite-react-forked-tgv2tj?workspaceId=ab98d85e-d21a-46f5-ab70-5be2562476c1
2. Navigate to the collections page
3. Navigate back
4.  Take heap snapshots
5. Repeat this a bunch of times

### Current behavior

In the Memory tab in the Chrome dev tools, taking snapshots (Both heap shots or detached nodes) after navigating to and back from both pages, detached nodes start accumulating.

### Expected behavior

Nodes should not become detached when navigating to and away from the page and should eventually be cleared from memory.

### Context

I am trying to use the component to display data, however, the detached nodes start to accumulate without being collected by the garbage collector.

### Your environment

<details>
  <summary><code>npx @mui/envinfo</code></summary>

```
System:
    OS: Linux 6.1 Debian GNU/Linux 12 (bookworm) 12 (bookworm)
  Binaries:
    Node: 20.9.0 - /usr/local/bin/node
    npm: 9.8.1 - /usr/local/bin/npm
    pnpm: 8.10.2 - /usr/local/share/npm-global/bin/pnpm
  Browsers:
    Chrome: Not Found
  npmPackages:
    @emotion/react: ^11.13.3 => 11.13.3 
    @emotion/styled: ^11.13.0 => 11.13.0 
    @mui/core-downloads-tracker:  6.1.7 
    @mui/material: ^6.1.7 => 6.1.7 
    @mui/private-theming:  6.1.7 
    @mui/styled-engine:  6.1.7 
    @mui/styled-engine-sc: ^6.1.7 => 6.1.7 
    @mui/system:  6.1.7 
    @mui/types:  7.2.19 
    @mui/utils:  6.1.7 
    @mui/x-data-grid: ^7.22.2 => 7.22.2 
    @mui/x-internals:  7.21.0 
    @types/react:  18.3.12 
    react: ^18.3.1 => 18.3.1 
    react-dom: ^18.3.1 => 18.3.1 
    styled-components: ^6.1.13 => 6.1.13 
</details>
```

**Search keywords**: memory leak Datagrid

**Comment by michelengelen** [comment]
@romgrk could you have a look at this one, please? 🙏🏼

**Comment by lauri865** [comment]
I can confirm this on our end. The parent container/component of the grid doesn't get garbage collected. But I wonder what it is – must be something that reaches the global scope (event handlers not getting removed properly, or something along the lines), as everything else should get cleaned up by React automatically when the whole page unmounts.

**Comment by polyamakov** [comment]
Any news on this? We are worried about this too.

**Comment by lauri865** [comment]
That seems to be the case, mostly at least. And explains why it's often the whole route's content that's being retained, but not always.

Interestingly, I also tried to forcefully remove the rootElement's content and rootElement itself from the DOM on unmount, which reduced the memory allocation for the reproduction at least. But for some odd reason, Datagrid's main container still was retained, which could signal that there's a potential issue somewhere 🤔.

On thing I noticed quickly is that the following timer is not cleared on unmount (but probably not the root cause of this issue):
https://github.com/mui/mui-x/blob/564be338215883a3d2f5fd33a852800d7eb03e05/packages/x-data-grid/src/hooks/features/dimensions/useGridDimensions.ts#L124-L127

Reproduction: https://codesandbox.io/p/devbox/vite-react-forked-7frhqz?workspaceId=8a8eeabe-fead-479a-a993-25a9868e8015

<img width="742" alt="Screenshot 2024-11-26 at 00 51 40" src="https://github.com/user-attachments/assets/79726161-f441-44ec-a65b-03182b9cc58f">
